extensions/security/deployment/src/main/java/io/quarkus/security/deployment/SecurityProcessor.java

```diff
@@ -85,6 +109,147 @@ void registerSecurityInterceptors(BuildProducer<InterceptorBindingRegistrarBuild
+                });
+                configurator.done();
+            }
+        }));
+    }
+
+    private void registerSecuredMethod(ResultHandle checkStorage,
+            MethodCreator methodCreator,
+            Map.Entry<MethodInfo, AnnotationInstance> methodEntry) {
+        try {
+            MethodInfo method = methodEntry.getKey();
+            ResultHandle aClass = methodCreator.load(method.declaringClass().name().toString());
+            ResultHandle methodName = methodCreator.load(method.name());
+            ResultHandle params = paramTypes(methodCreator, method.parameters());
+
+            AnnotationInstance instance = methodEntry.getValue();
+            ResultHandle securityAnnotation = methodCreator.load(instance.name().toString());
+
+            ResultHandle annotationParameters = annotationValues(methodCreator, instance);
+
+            Method registerAnnotation = SecurityCheckStorageBuilder.class.getDeclaredMethod("registerAnnotation",
+                    String.class, String.class, String[].class, String.class, String[].class);
+            methodCreator.invokeVirtualMethod(MethodDescriptor.ofMethod(registerAnnotation), checkStorage,
+                    aClass, methodName, params, securityAnnotation, annotationParameters);
+        } catch (NoSuchMethodException e) {
+            throw new IllegalStateException("registerAnnotation method not found on on SecurityCheckStorage", e);
+        }
+    }
+
+    private ResultHandle annotationValues(MethodCreator methodCreator, AnnotationInstance instance) {
+        AnnotationValue value = instance.value();
+        if (value != null && value.asStringArray() != null) {
+            String[] values = value.asStringArray();
+            ResultHandle result = methodCreator.newArray(String.class, methodCreator.load(values.length));
+            int i = 0;
+            for (String val : values) {
+                methodCreator.writeArrayValue(result, i, methodCreator.load(val));
+            }
+            return result;
+        }
+        return methodCreator.loadNull();
+    }
+
+    private ResultHandle paramTypes(MethodCreator ctor, List<Type> parameters) {
+        ResultHandle result = ctor.newArray(String.class, ctor.load(parameters.size()));
+
+        for (int i = 0; i < parameters.size(); i++) {
+            ctor.writeArrayValue(result, i, ctor.load(parameters.get(i).toString()));
+        }
+
+        return result;
+    }
+
+    private Map<MethodInfo, AnnotationInstance> gatherSecurityAnnotations(ApplicationIndexBuildItem indexBuildItem,
+            BeanRegistrar.RegistrationContext registrationContext) {
+        Set<DotName> securityAnnotations = SecurityAnnotationsRegistrar.SECURITY_BINDINGS.keySet();
+        AnnotationStore annotationStore = registrationContext.get(BuildExtension.Key.ANNOTATION_STORE);
+        Set<ClassInfo> classesWithSecurity = new HashSet<>();
+
+        Collection<ClassInfo> classes = indexBuildItem.getIndex().getKnownClasses();
```

Comment:
Does this really need to iterate over all the classes? Can't you just look for security annotations in the index?

Reply:
The problem with using the index is with the synthetic annotations (like the `@DenyAll` when the specific switches are enabled or the `@Secured` transalted to `@RolesAllowed` that @aureamunoz is working on).
I was trying to work around it by emitting build items that would contain such information. But it led to build step cycle with resteasy. I think it was because I use ResteasyDeployment for denying JAX-RS by default.
I'm not 100% sure if it cannot be worked around but thought going through the application index only shouldn't be that costly.

Reply:
Plus annotationStore is populated lazily. That's why I could not get a list of classes with synthetic annotations from it.

Reply:
I haven't followed this PR lately, but from the start I thought that the BuildItem approach would be the cleanest way to go. 
I'll look into the PR more tomorrow and see if / how it can be done

Reply:
@geoand suggested looping through the index only if there are synthetically added annotations (only if the config options that result in adding `@DenyAll`'s are present).
The newest patchset does it. 
WDYT @stuartwdouglas ?

Reply:
The reason we need to loop through the index when denyAll is enabled is because like @michalszynkiewicz mentione above, the annotation store doesn't provide a way to give all the instances of an annotation in the way the Jandex index does.
And the implementation of denyAll is based on the annotation store, since it's the easient way to do it and be sure that `@DenyAll` is only added to CDI beans an not everything under the sun.